Fix silent error in EVP_CIPHER_CTX_get_updated_iv.

nblaismiko opened the pull request:
Fixes [#27117](https://github.com/openssl/openssl/issues/27117)

    Fix silent error in EVP_CIPHER_CTX_get_updated_iv.

    Added new params API function OSSL_PARAM_set_octet_string_or_ptr to only
    call the correct setter for OSSL_CIPHER_PARAM_IV and OSSL_CIPHER_PARAM_UPDATED_IV.
    Both OSSL_PARAM_set_octet_string and OSSL_PARAM_set_octet_ptr could be called with
    only one expected to succeed. This would put a silent error on the error stack when
    calling EVP_CIPHER_CTX_get_updated_iv.

t8m commented:
Unfortunately this is not acceptable with CLA: trivial. Could you please sign a CLA?

nblaismiko commented:
I don't mind to sign a CLA, looking into what I need to do for that.

t8m commented:
@paulidale When you are adding a hold could you please explain more concretely what do you expect to be added to the PR?

In particular here I do not really see how the `hold: needs documentation` applies.

nblaismiko commented:
Sent the signed ICLA agreement to cla@openssl.org.

Also fixed the format validation errors in added code and changed the email of the commits to my personal address.

t8m commented:
I've removed the holds for tests (you've added it) and docs (I do not see any required.)

nblaismiko commented:
Coding style validation failing on code I haven't changed, should I go ahead and fix it or leave it as is? Sorry if it's an obvious question, prefer to ask rather than require another pipeline run.

```
Style report for commits 131fff1b09e07eeb5db8b99d7e8f502d8c4fb1e5..dfedc87c2e255f8e00627cb982663fa07c9bb810
  providers/implementations/ciphers/ciphercommon.c:176:indent = 8 != 55 for hanging '(':        ossl_unused void *cctx, ossl_unused void *provctx
  providers/implementations/ciphers/ciphercommon.c:177:indent = 4 != 55 for hanging '(':    )
```

t8m commented:
> Coding style validation failing on code I haven't changed, should I go ahead and fix it or leave it as is? Sorry if it's an obvious question, prefer to ask rather than require another pipeline run.
> 
> ```
> 
> ```

Just ignore it, if it is in unrelated part of the file.

openssl-machine commented:
24 hours has passed since 'approval: done' was set, but this PR has failing CI tests. Once the tests pass it will get moved to 'approval: ready to merge' automatically,alternatively please review and set the label manually.

slontis reviewed (commented):
Would it be better to do this as an inline function added to params.h

````
static ossl_unused ossl_inline int OSSL_PARAM_set_octet_string_or_ptr(
OSSL_PARAM *p, const void *val, size_t used_len)
{
    if (p != NULL) {
        int res = 0;

        if (p->data_type == OSSL_PARAM_OCTET_STRING)
            res = OSSL_PARAM_set_octet_string(p, *val, len);
        else if (p->data_type == OSSL_PARAM_OCTET_PTR)
            res = OSSL_PARAM_set_octet_ptr(p, val, len);
        if (res == 0) {
            ERR_raise(ERR_LIB_PROV, PROV_R_FAILED_TO_SET_PARAMETER);
            return 0;
        }
    }
    return 1;
}
````

paulidale commented:
> Would it be better to do this as an inline function added to params.h


I wouldn't.  It is not avoiding a function call and will bloat code needlessly.

slontis commented:
I think they are still useful in some places.. They are used to pass config items to the providers for example.

paulidale commented:
> They are used to pass config items to the providers for example.

This is the **only** use case we have where they make sense and even then, they could be done using a UTF8_STRING or OCTET_STRING without any loss.

nblaismiko commented:
> Would it be better to do this as an inline function added to params.h

I agree it does look cleaner, I had not though of it. I've had a feeling while working on this that params being either an octet_string or an octer_ptr is something that is being phased out or discouraged. Adding a specific setter for that case may make this approach look favorable.

I'm new to this codebase so I don't mind doing it in any way that more experienced contributors and maintainers prefer.

slontis commented:
I would see if other reviewers have opinions on this before changing anything..

Pauli - I dont see us removing the ptr variant any time soon, due to backwards compatibility.

paulidale commented:
We can't remove it but we could stop using it & flag it as deprecated.

nblaismiko commented:
Made the suggested adjustment, I can always go back if needed. 

Not sure why we passed the pointer to the IV array with & for PTR since it's cast to void* anyways, so I removed it. Let me know if I missed something with that.

I should be added to the CLA database now.

mattcaswell commented on include/openssl/params.h:
```diff
@@ -157,6 +157,18 @@ OSSL_PARAM *OSSL_PARAM_dup(const OSSL_PARAM *p);
 void OSSL_PARAM_free(OSSL_PARAM *p);
 
+static ossl_unused ossl_inline int
+OSSL_PARAM_set_octet_string_or_ptr(OSSL_PARAM *p, const void *val, size_t used_len)
```
If you are adding a new API function you need to also add documentation for it (e.g. see doc/man3/OSSL_PARAM_int.pod)

nblaismiko replied:
I took a crack at adding the documentation, hopefully the way I did it makes sense.

I also removed the inline as it was making the API and documentation feel messy. I don't think the optimization is worth it here, but I can re-add it if it's important.

nblaismiko commented:
Fixed conflict.

mattcaswell commented:
Pushed to master. Thank you for your contribution.
